web3.js/src/connection.ts

```diff
@@ -4281,75 +4240,248 @@ export class Connection {
+                if (
+                  this._subscriptionsAutoDisposedByRpc.has(serverSubscriptionId)
+                ) {
+                  /**
+                   * Special case.
+                   * If we're dealing with a subscription that has been auto-
+                   * disposed by the RPC, then we can skip the RPC call to
+                   * tear down the subscription here.
+                   *
+                   * NOTE: There is a proposal to eliminate this special case, here:
+                   * https://github.com/solana-labs/solana/issues/18892
+                   */
+                  this._subscriptionsAutoDisposedByRpc.delete(
+                    serverSubscriptionId,
+                  );
+                } else {
+                  this._subscriptionsByHash[hash] = {
+                    ...subscription,
+                    state: 'unsubscribing',
+                  };
+                  try {
+                    await this._rpcWebSocket.call(unsubscribeMethod, [
+                      serverSubscriptionId,
+                    ]);
+                  } catch (e) {
+                    if (e instanceof Error) {
+                      console.error(`${unsubscribeMethod} error:`, e.message);
+                    }
+                    if (!isCurrentConnectionStillActive()) {
+                      return;
+                    }
+                    // TODO: Maybe add an 'errored' state or a retry limit?
+                    this._subscriptionsByHash[hash] = {
+                      ...subscription,
+                      state: 'subscribed',
+                    };
+                    await this._updateSubscriptions();
+                    return;
+                  }
+                }
+                this._subscriptionsByHash[hash] = {
+                  ...subscription,
+                  state: 'unsubscribed',
+                };
+                await this._updateSubscriptions();
+              })();
+            }
+            break;
+          case 'subscribing':
+          case 'unsubscribing':
+            break;
+        }
+      }),
+    );
+  }
 
-    for (let id of rootKeys) {
-      const sub = this._rootSubscriptions[id];
-      this._subscribe(sub, 'rootSubscribe', []);
+  /**
+   *
+   * @internal
+   */
+  _handleServerNotification<TCallback extends SubscriptionConfig['callback']>(
+    serverSubscriptionId: ServerSubscriptionId,
+    callbackArgs: Parameters<TCallback>,
+  ): void {
+    const callbacks =
+      this._subscriptionCallbacksByServerSubscriptionId[serverSubscriptionId];
+    if (callbacks === undefined) {
+      return;
     }
-
-    for (let id of logsKeys) {
-      const sub = this._logsSubscriptions[id];
-      let filter;
-      if (typeof sub.filter === 'object') {
-        filter = {mentions: [sub.filter.toString()]};
-      } else {
-        filter = sub.filter;
+    callbacks.forEach(cb => {
+      try {
+        cb(...(callbackArgs as [any, any]));
+      } catch (e) {
```

Comment:
This is a behavior change but probably a good one. I'm guessing you're trying to help out app developers here in case their callback throws an error?

Reply:
Yeah. I was _shocked_ that we've gotten away this long without wrapping user-supplied callbacks in guards. Right now, one bad callback can bring down the whole publish queue.